lib/Dialect/TTNN/Transforms/TTNNWorkarounds.cpp

```diff
@@ -0,0 +1,433 @@
+    llvm::outs() << "Memory layout workaround change\n";
+    modified =
+        tensorMemoryLayout != *workaround.getTensorMemoryLayoutWorkaround()
+                                   .getTensorMemoryLayoutWorkaround();
+    tensorMemoryLayout = *workaround.getTensorMemoryLayoutWorkaround()
+                              .getTensorMemoryLayoutWorkaround();
+  }
+
+  return modified;
+}
+
+// Helper method to propagate output changes into DPS inits. It accepts output
+// result as input argument. It iterates over all the uses of the output result
+// and updates the DPS init with the new output type.
+static void propagateOutputChangesIntoDPSInits(OpResult &outputResult) {
+  // Iterate over all the uses of the outputResult
+  for (auto &use : outputResult.getUses()) {
+    // Get the user operation
+    Operation *userOp = use.getOwner();
+
+    // Check if the user operation is a DPS op
+    if (mlir::isa<DestinationStyleOpInterface>(userOp)) {
+      DestinationStyleOpInterface dpsOp =
+          mlir::dyn_cast<DestinationStyleOpInterface>(userOp);
+
+      // Check if the use is a DPS init
+      if (dpsOp && dpsOp.isDpsInit(&use)) {
+        // Update the DPS init with the new output type
+        OperandRange dpsInits = dpsOp.getDpsInits();
+        dpsOp
+            ->getResult(dpsInits.getBeginOperandIndex() -
+                        use.getOperandNumber())
+            .setType(outputResult.getType());
+      }
+    }
+  }
+}
+
+// Helper method to apply input operand workarounds. It accepts inputOperand,
+// workaround, rewriter and current operation as input arguments. It applies the
+// workarounds on the input operand and returns true if the workarounds were
+// applied.
+static bool workaroundInputOperand(OpOperand &inputOperand,
+                                   TTNNOperandWorkarounds &workaround,
+                                   PatternRewriter &rewriter,
+                                   TTNNWorkaroundInterface op) {
+  bool modified = false;
+  // Get the input operand type to extract the tensor layout, buffer type and
+  // memory layout
+  auto inputOperandType =
+      mlir::cast<RankedTensorType>(inputOperand.get().getType());
+  ::mlir::tt::ttnn::TTNNLayoutAttr inputLayoutAttr =
+      mlir::cast<::mlir::tt::ttnn::TTNNLayoutAttr>(
+          inputOperandType.getEncoding());
+  Layout tensorLayout =
+      llvm::isa<TileType>(inputLayoutAttr.getMemref().getElementType())
+          ? Layout::Tile
+          : Layout::RowMajor;
+  BufferType tensorBufferType = inputLayoutAttr.getBufferType();
+  TensorMemoryLayout tensorMemoryLayout = inputLayoutAttr.getMemLayout();
+
+  // Apply the workarounds on the input operand workadound arguments
+  modified |= applyTensorLayoutWorkaround(workaround, tensorLayout);
+  modified |= applyTensorBufferTypeWorkaround(workaround, tensorBufferType);
+  modified |= applyTensorMemoryLayoutWorkaround(workaround, tensorMemoryLayout);
```

Comment:
I don't think this is a good pattern - the user of the function shouldn't have to supply the current state (like `tensorLayout`), it's error prone.

Can this be redesigned so that the `applyTensorLayoutWorkaround` takes in an op (or operand) and the workaround?

On another note, it'd be great (if possible!) to have an `applyWorkaround<T>(operand, T workaround)`. And then on top of that, a function that takes in a variadic list of different workarounds (maybe std::variant could help) that would then, based on type of each workaround, call corresponding `applyWorkaround<T>(operand, T workaround)`.

@rpavlovicTT you're good with meta-programming, any suggestions?